go/src/koding/klient/machine/mount/sync/history/history.go

```diff
@@ -0,0 +1,161 @@
+package history
+
+import (
+	"container/ring"
+	"fmt"
+	"sync"
+	"time"
+
+	msync "koding/klient/machine/mount/sync"
+)
+
+// Record stores a single history record.
+type Record struct {
+	CreatedAt time.Time `json:"createdAt"`        // creation time.
+	Message   string    `json:"message"`          // short summary.
+	Details   string    `json:"detail,omitempty"` // detailed description.
+}
+
+// History gathers synchronization history of results produced by stored Syncer.
+type History struct {
+	s msync.Syncer // underlying Syncer.
+
+	mu sync.Mutex
+	r  *ring.Ring
+
+	once  sync.Once
+	stopC chan struct{} // channel used to close any opened exec streams.
+}
+
+// NewHistory creates a new History instance. Provided size defines the maximum
+// length of history records.
+func NewHistory(s msync.Syncer, size int) *History {
+	return &History{
+		s:     s,
+		r:     ring.New(size),
+		stopC: make(chan struct{}),
+	}
+}
+
+// ExecStream wraps underlying execers with history gathering logic.
+func (h *History) ExecStream(evC <-chan *msync.Event) <-chan msync.Execer {
+	exhC := make(chan msync.Execer)
+
+	go func() {
+		defer close(exhC)
+
+		exC := h.s.ExecStream(evC)
+		for {
+			select {
+			case ex, ok := <-exC:
+				if !ok {
+					return
+				}
+
+				h.add(&Record{
+					CreatedAt: time.Now(),
```

Comment:
prob better to use UTC, time might change on day light saving ( or similar ) cases